plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/LayerConstraintPostprocessor.java

```diff
@@ -0,0 +1,246 @@
+        
+        List<Layer> layers = layeredGraph.getLayers();
+        
+        // If the list of layers is non-empty, we may have FIRST and LAST nodes to move around
+        if (!layers.isEmpty()) {
+            // Retrieve the current first and last layers
+            Layer firstLayer = layers.get(0);
+            Layer lastLayer = layers.get(layers.size() - 1);
+            
+            // We may also need label dummy layers between the very first / last layers and the first / last layers
+            Layer firstLabelLayer = new Layer(layeredGraph);
+            Layer lastLabelLayer = new Layer(layeredGraph);
+
+            // Move our nodes!
+            moveFirstAndLastNodes(layeredGraph, firstLayer, lastLayer, firstLabelLayer, lastLabelLayer);
+            
+            // Add non-empty label layers
+            if (!firstLabelLayer.getNodes().isEmpty()) {
+                layers.add(0, firstLabelLayer);
+            }
+            
+            if (!lastLabelLayer.getNodes().isEmpty()) {
+                layers.add(lastLabelLayer);
+            }
+        }
+        
+        // Restore hidden FIRST_SEPARATE and LAST_SEPARATE nodes, if any
+        if (layeredGraph.hasProperty(InternalProperties.HIDDEN_NODES)) {
+            // Create the new first and last layers, in case they will be needed
+            Layer firstSeparateLayer = new Layer(layeredGraph);
+            Layer lastSeparateLayer = new Layer(layeredGraph);
+            
+            // Restore our nodes
+            restoreHiddenNodes(layeredGraph, firstSeparateLayer, lastSeparateLayer);
+            
+            // Add non-empty separate layers
+            if (!firstSeparateLayer.getNodes().isEmpty()) {
+                layers.add(0, firstSeparateLayer);
+            }
+            
+            if (!lastSeparateLayer.getNodes().isEmpty()) {
+                layers.add(lastSeparateLayer);
+            }
+        }
+        
+        monitor.done();
+    }
+
+    /**
+     * Moves nodes with {@link LayerConstraint#FIRST} and {@link LayerConstraint#LAST} to their respective layers and
+     * removes layers that have become empty.
+     */
+    private void moveFirstAndLastNodes(final LGraph layeredGraph, final Layer firstLayer, final Layer lastLayer,
+            final Layer firstLabelLayer, final Layer lastLabelLayer) {
+        
+        // We'll start by moving FIRST and LAST nodes (as well as connected label dummies) to their proper layers
+        for (Layer layer : layeredGraph) {
+            // Iterate through a node array to avoid ConcurrentModificationExceptions
+            LNode [] nodes = LGraphUtil.toNodeArray(layer.getNodes());
+            
+            for (LNode node : nodes) {
+                switch (node.getProperty(LayeredOptions.LAYERING_LAYER_CONSTRAINT)) {
+                case FIRST:
+                    node.setLayer(firstLayer);
+                    throwUpUnlessNoIncomingEdges(node);
```

Comment:
Check this _before_ moving the node? Not that it would make much of a difference ...

Reply:
Yeah, but it feels a little cleaner. Will do.